examples/md-flexible/src/domainDecomposition/RegularGridDecomposition.cpp

```diff
@@ -275,31 +278,128 @@ void RegularGridDecomposition::reflectParticlesAtBoundaries(AutoPasType &autoPas
-              isUpper ? reflSkinMax[dimensionIndex] + (reflSkinMax[dimensionIndex] - position[dimensionIndex])
-                      : reflSkinMin[dimensionIndex] - position[dimensionIndex];
-          mirrorParticle.setR(position);
+        const auto distanceToBoundary = std::abs(position[dimensionIndex] - boundaryPosition);
+
+        // For single-site molecules, we discard molecules further than sixthRootOfTwo * sigma, and are left only with
+        // molecules who will experience repulsion from the boundary.
+
+        // For multi-site molecules, we discard molecules with center-of-mass further than sixthRootOfTwo * the largest
+        // sigma of any site of that molecule. Some molecules may still experience attraction in which case their force
+        // is reset to before the force interaction.
+        //
+        // Note, there is a scenario where a molecule has center-of-mass further than sixthRootOfTwo * sigma, but has a
+        // site closer than this distance, with a large enough epsilon, that repulsion would occur. For computational
+        // cost reasons, this scenario is neglected - no repulsion occurs. This *should*, in theory, with an appropriate
+        // step-size and molecular model, not cause any problems.
+
+        // To produce a "mirrored" multi-site molecule that could be used with the multi-site molecule functor would be
+        // very nasty. As such, we reimplement the kernel of the lennard-jones force here. We also use this for
+        // single-site molecules, primarily for consistency but it is also suspected to be cheaper than creating a
+        // mirror particle for use with the actual functor.
+
+        // Calculates force acting on site from another site
+        const auto LJKernel = [](const std::array<double, 3> sitePosition,
+                                 const std::array<double, 3> mirrorSitePosition, const double sigmaSquared,
+                                 const double epsilon24) {
+          const auto displacement = autopas::utils::ArrayMath::sub(sitePosition, mirrorSitePosition);
+          const auto distanceSquared = autopas::utils::ArrayMath::dot(displacement, displacement);
+
+          const auto inverseDistanceSquared = 1. / distanceSquared;
+          const auto lj2 = sigmaSquared * inverseDistanceSquared;
+          const auto lj6 = lj2 * lj2 * lj2;
+          const auto lj12 = lj6 * lj6;
+          const auto lj12m6 = lj12 - lj6;
+          const auto scalarMultiple = epsilon24 * (lj12 + lj12m6) * inverseDistanceSquared;
+
+          return autopas::utils::ArrayMath::mulScalar(displacement, scalarMultiple);
+        };
+
+        const bool reflectMoleculeFlag =
+#if MD_FLEXIBLE_MODE == MULTISITE
+            distanceToBoundary < sixthRootOfTwo * particlePropertiesLib.getMoleculesLargestSigma(p->getTypeId()) / 2.;
+#else
+            distanceToBoundary < sixthRootOfTwo * particlePropertiesLib.getSigma(p->getTypeId()) / 2.;
+#endif
```

Comment:
```suggestion
            distanceToBoundary < sixthRootOfTwo * 0.5 *
#if MD_FLEXIBLE_MODE == MULTISITE
              particlePropertiesLib.getMoleculesLargestSigma(p->getTypeId());
#else
              particlePropertiesLib.getSigma(p->getTypeId());
#endif
```
(not sure about indentation :sweat_smile: )